### PR TITLE
Added initial key value API

### PR DESCRIPTION
**Pull request by codyebberson**
See https://github.com/medplum/medplum/discussions/3995

TODO:
- [x] Cap users at 10 key/value pairs
- [x] Add utility methods to `MedplumClient`

### CODE REVIEW

File: packages/core/src/client.ts [hunk 1]
```diff
@@ -2712,6 +2714,17 @@ export class MedplumClient extends EventTarget {
     return this.request('POST', url, options);
   }
 
+  /**
+   * Returns the key value client.
+   * @returns The key value client.
+   */
+  get keyValue(): MedplumKeyValueClient {
```
Comment (ThatOneBro) [inline packages/core/src/client.ts:2721]: Love to see it. Just the right amount of magic here.

I think we can also lazy load the `./keyvalue` module here but not blocking this PR.

We should definitely do this to the `FHIRcast` client and `SubscriptionManager` too I think. If we actually bundle them as (a) separate file(s) potentially could reduce bundle size and initial load times a lot for those who are not actually using any of those fancy features.

File: packages/server/src/keyvalue/store.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  }
+  if (value.length > MAX_VALUE_LENGTH) {
+    throw new OperationOutcomeError(badRequest('Value too long'));
+  }
+  const ctx = getAuthenticatedContext();
+  const cacheKey = getCacheKey(ctx);
+  const redis = getRedis();
+  const exists = await redis.hexists(cacheKey, key);
+  if (!exists) {
+    const length = await redis.hlen(cacheKey);
```
Comment (ThatOneBro) [inline packages/server/src/keyvalue/store.ts:42]: I see the switch to using raw Redis in tests may have paid off here

File: packages/server/src/keyvalue/store.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+/**
+ * Returns the Redis cache key for the given project and key.
+ * Validates the key string is an alphanumeric string from 1-255 characters.
+ * @param ctx - The request context.
+ * @returns The Redis cache key.
+ */
+function getCacheKey(ctx: AuthenticatedRequestContext): string {
+  return `medplum:keyvalue:${ctx.membership.id}`;
```
Comment (ThatOneBro) [inline packages/server/src/keyvalue/store.ts:82]: Is this per project or per project per user here? Should it not be `ctx.project.id`? Or are these KVs per user? If they are per user, is there any way to share keys for a whole project with this API?

**Review by ThatOneBro — COMMENTED** [review]
Looks good for the most part, one major clarification to make sure I understand what the scoping is for this system

File: packages/server/src/keyvalue/store.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    throw new OperationOutcomeError(badRequest('Invalid key'));
+  }
+  if (!/^[a-zA-Z0-9_-]+$/.test(key)) {
+    throw new OperationOutcomeError(badRequest('Invalid key'));
+  }
+  return key;
+}
+
+/**
+ * Returns the Redis cache key for the given project and key.
```
Comment (ThatOneBro) [inline packages/server/src/keyvalue/store.ts:76]: Mostly I think I was confused by the description here: `for the given project and key`

File: packages/server/src/keyvalue/store.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  const cacheKey = getCacheKey(ctx);
+  const redis = getRedis();
+  const exists = await redis.hexists(cacheKey, key);
+  if (!exists) {
+    const length = await redis.hlen(cacheKey);
+    if (length >= MAX_ITEMS) {
+      throw new OperationOutcomeError(badRequest('Max items exceeded'));
+    }
+  }
+  await redis.hset(cacheKey, key, value);
```
Comment (mattlong) [inline packages/server/src/keyvalue/store.ts:47]: There's a small race condition here where someone could spam a lot of set requests at the same time and have more than `MAX_ITEMS` wind up being set. We could avoid it by obtaining/releasing a lock around the operation, but I'm not sure if it's worth the complexity.

File: packages/server/src/keyvalue/store.ts (hunk 3 above)
Comment (codyebberson) [reply packages/server/src/keyvalue/store.ts:82]: It is currently per user, by design.  My thinking was that the primary use case for this is temp storage for bots, not for any data sharing.  I feel like the moment we open it up to sharing, it really spirals in complexity (access policies, hidden values, readonly values, auditing, logging, etc).  Purely per user avoids all of that complexity.

File: packages/server/src/keyvalue/store.ts (hunk 4 above)
Comment (codyebberson) [reply packages/server/src/keyvalue/store.ts:76]: Ah, good catch, that comment is outdated.  I will update.

File: packages/server/src/keyvalue/store.ts (hunk 5 above)
Comment (codyebberson) [reply packages/server/src/keyvalue/store.ts:47]: Acknowledged.  If someone gets 11 (or even 20) keys, that's not going to kill us.

**Review by ThatOneBro — APPROVED** [review]

**Review by mattlong — APPROVED** [review]

**Pull request merged**